lib/IDE/Utils.cpp

```diff
@@ -1181,3 +1181,75 @@ std::pair<Type, ConcreteDeclRef> swift::ide::getReferencedDecl(Expr *expr) {
 
   return std::make_pair(exprTy, refDecl);
 }
+
+bool swift::ide::isBeingCalled(ArrayRef<Expr *> ExprStack) {
+  if (ExprStack.empty())
+    return false;
+
+  Expr *Target = ExprStack.back();
+  auto UnderlyingDecl = getReferencedDecl(Target).second;
+  for (Expr *E: reverse(ExprStack)) {
+    auto *AE = dyn_cast<ApplyExpr>(E);
+    if (!AE || AE->isImplicit())
+      continue;
+    if (isa<ConstructorRefCallExpr>(AE) && AE->getArg() == Target)
+      return true;
+    if (isa<SelfApplyExpr>(AE))
+      continue;
+    if (getReferencedDecl(AE->getFn()).second == UnderlyingDecl)
+      return true;
+  }
+  return false;
+}
+
+static Expr *getContainingExpr(ArrayRef<Expr *> ExprStack, size_t index) {
+  if (ExprStack.size() > index)
+    return ExprStack.end()[-std::ptrdiff_t(index + 1)];
+  return nullptr;
+}
+
+Expr *swift::ide::getBase(ArrayRef<Expr *> ExprStack) {
+  if (ExprStack.empty())
+    return nullptr;
+
+  Expr *CurrentE = ExprStack.back();
+  Expr *ParentE = getContainingExpr(ExprStack, 1);
+  if (auto DSE = dyn_cast_or_null<DotSyntaxCallExpr>(ParentE))
+    return DSE->getBase();
+  else if (auto MRE = dyn_cast<MemberRefExpr>(CurrentE))
+    return MRE->getBase();
+  else if (auto SE = dyn_cast<SubscriptExpr>(CurrentE))
+    return SE->getBase();
+  return nullptr;
+}
+
+bool swift::ide::isDynamicCall(Expr *Base, ValueDecl *D) {
+  auto TyD = D->getDeclContext()->getSelfNominalTypeDecl();
+  if (!TyD)
+    return false;
+
+  if (isa<StructDecl>(TyD) || isa<EnumDecl>(TyD) || D->isFinal())
+    return false;
+
+  // super.method(), but is dynamic if the decl is marked as dynamic
+  if (Base->isSuperExpr() && !D->isDynamic())
```

Comment:
This used to use isSuperRefExpr, which also checked for a LoadExpr (isSuperExpr does not). Removing it didn't cause any tests to fail... anyone know what case that would have been for?

Reply:
I think we should dig into this a bit:
* Construct a test case to see what can happen here. Is the "injection" possible in pure Swift code or is this only for @objc classes?
* If we are going to mark a `super` call as dynamic we need to distinguish this case from other dynamic calls, otherwise this may get handled as the other dynamic calls that may go to override classes.

Without additional information to distinguish the cases I would argue it is better to treat it as "non-dynamic" since in the vast majority of cases it will do the right thing and go to the one function in the super class, instead of treating it as override methods may get called which is actually always wrong.

Reply:
It's only possible with @objc and dynamic. You're right that we probably want to distinguish this separately, including in the index. I'll leave this to another PR.